### DISCUSSION TITLE
"A transaction is already begun on this Session" and "MissingGreenlet" errors with AsyncSession

### DISCUSSION THREAD

**Discussion by AnasAitzouinet** [sqlalchemy/sqlalchemy#12140] [Usage Questions]
Hello,

I'm working on a FastAPI application using SQLAlchemy 2.0's AsyncSession with PostgreSQL, and I'm encountering some issues related to transaction management. I have a custom transaction context manager that I need to use throughout my application.

Environment:

Python version: 3.x
SQLAlchemy version: 2.0.20
FastAPI version: 0.98.0
asyncpg version: 0.27.0
Database: PostgreSQL
Problem Description:

I'm trying to use a custom transaction context manager with AsyncSession, but I'm facing the following errors:
 
1. InvalidRequestError: A transaction is already begun on this Session.
2. sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_only() here. Was IO attempted in an unexpected place?

I want to understand why these errors are occurring and how to resolve them while keeping my custom transaction context manager.

**Code Snippets:**

**Session Setup (session.py):**

```python
from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession
from sqlalchemy.orm import sessionmaker
from contextlib import asynccontextmanager
import logging

logger = logging.getLogger(__name__)

engine = create_async_engine(
    settings.DATABASE_URL,
    echo=True,
    future=True,  # Using SQLAlchemy 2.0 style
)

SessionLocal = sessionmaker(
    bind=engine,
    class_=AsyncSession,
    expire_on_commit=True,
    autoflush=True,
    autobegin=True,  # I intend to manage transactions explicitly
)

async def get_db():
    async with SessionLocal() as session:
        yield session

@asynccontextmanager
async def transaction(db: AsyncSession):
    if not db.in_transaction():
        async with db.begin():
            logger.debug("Explicit transaction begin")
            yield
        logger.debug("Explicit transaction commit")
    else:
        logger.debug("Already in transaction")
        yield
        if db.in_transaction():
            await db.commit()
            logger.debug("Implicit transaction commit")

```

**Service Method (start_time_entry in time_entry_service.py):**

```python
class TimeEntryService:
    def __init__(self, db: AsyncSession, user: User):
        self.db = db
        self.user = user

    async def start_time_entry(self, request: TimeEntryCreate):
        async with transaction(self.db):
            open_entry = await self.check_active_entry()

            if open_entry:
                raise HTTPException(status_code=400, detail="An active time entry already exists.")

            new_entry = TimeEntry(
                user_id=self.user.id,
                project_id=request.project_id,
                start_time=datetime.now(timezone.utc),
                status=TimeEntryStatusEnum.open,
                description=request.description or None,
            )
            self.db.add(new_entry)
            await self.db.commit()
            await self.db.refresh(new_entry)
            return new_entry
``` 

** Error Messages **
```
sqlalchemy.exc.InvalidRequestError: A transaction is already begun on this Session.
``` 

```
sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_only() here. Was IO attempted in an unexpected place? 
```

### What I've Tried:

1. Adjusting session configurations (autobegin, autocommit, autoflush, expire_on_commit).
2. Removing manual commit() calls within the transaction context.
3. Modifying the custom transaction context manager to handle transactions differently.

**My Goal:**

1. Use a custom transaction context manager: I need to maintain this pattern due to the structure of my application and for consistency across different services.
2. Manage transactions explicitly: I prefer to have explicit control over when transactions begin and end.

**Questions:**

1. Why am I encountering these errors when using a custom transaction context manager with AsyncSession?
2. How can I properly manage transactions in this context while still using my custom transaction manager?
3. Is there a recommended way to use a custom transaction context manager with AsyncSession in SQLAlchemy 2.0?

**Additional Context:**

1. The application ensures that only one AsyncSession is used per request.
2. I observed that removing the db.in_transaction() check changes the behavior but I want to keep it for cases where a transaction might already be active.
3. I suspect that manual commit() calls within the transaction block might be causing conflicts, but I'm unsure how to adjust the code accordingly.

Any guidance or suggestions would be greatly appreciated!

Thank you for your time.

**Comment by zzzeek** [comment]
These errors mean what they say, and without a stack trace it's impossible for me to tell you what the actual originating point is

>    InvalidRequestError: A transaction is already begun on this Session.

this means you called `session.begin()` when the session was already begun in a transaction, either via a previous call to `session.begin()`, or via autobegin, which [you can disable if you want to ensure you are in complete control of when the session begins](https://docs.sqlalchemy.org/en/20/orm/session_basics.html#session-autobegin-disable)

>    sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_only() here. Was IO attempted in an unexpected place?

this is when implicit IO occurs and there is a detailed guide on how this error happens and how to prevent it at https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#preventing-implicit-io-when-using-asyncsession .   Impossible to say what the originating cause is without a stack trace


overall I think your context manager needs some work here:

```
async def transaction(db: AsyncSession):
    if not db.in_transaction():
        async with db.begin():
            logger.debug("Explicit transaction begin")
            yield
        logger.debug("Explicit transaction commit")
    else:
        logger.debug("Already in transaction")
        yield
        if db.in_transaction():
            await db.commit()
            logger.debug("Implicit transaction commit")
```

this decorator doesnt make much sense, I can sort of see how the test for `in_transaction` and then the block for `with db.begin()` is meant to work, OK that means you want the outermost block to be the transaction, fine.  But then if you call the block **within** an existing transaction, it tries to....*end the transaction*?  Why would you want to do that?  it's not clear what you're going for here.

A typical way to do the kind of decorator this sort of looks like it wants to be would be, where there is a begin/commit at the outermost block only, would be the one we document at https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#session-subtransaction-behavior-removed, do just the begin block, that's it
